File: src/main/java/com/gtladd/gtladditions/common/machine/muiltblock/controller/AdvancedSpaceElevatorModuleMachine.java
package com.gtladd.gtladditions.common.machine.muiltblock.controller;

import org.gtlcore.gtlcore.api.machine.multiblock.ParallelMachine;
import org.gtlcore.gtlcore.common.data.GTLBlocks;
import org.gtlcore.gtlcore.common.data.machines.AdvancedMultiBlockMachine;
import org.gtlcore.gtlcore.common.machine.multiblock.electric.SpaceElevatorMachine;

import com.gregtechceu.gtceu.api.capability.GTCapabilityHelper;
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity;
import com.gregtechceu.gtceu.api.machine.feature.IRecipeLogicMachine;
import com.gregtechceu.gtceu.api.machine.multiblock.WorkableElectricMultiblockMachine;
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic;
import com.gregtechceu.gtceu.api.recipe.GTRecipe;
import com.gregtechceu.gtceu.utils.FormattingUtil;

import net.minecraft.ChatFormatting;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.world.level.Level;

import com.gtladd.gtladditions.api.recipeslogic.GTLAddMultipleRecipesLogic;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class AdvancedSpaceElevatorModuleMachine extends WorkableElectricMultiblockMachine implements ParallelMachine {

    private int SpaceElevatorTier = 0;
    private int ModuleTier = 0;
    private final boolean SEPMTier;

    public AdvancedSpaceElevatorModuleMachine(IMachineBlockEntity holder, boolean SEPMTier) {
        super(holder);
        this.SEPMTier = SEPMTier;
    }

    public @NotNull RecipeLogic createRecipeLogic(@NotNull Object @NotNull... args) {
        return new GTLAddMultipleRecipesLogic(this);
    }

    private void getSpaceElevatorTier() {
        Level level = this.getLevel();
        BlockPos pos = this.getPos();
        BlockPos[] coordinates = new BlockPos[] {
                pos.offset(8, -2, 3),
                pos.offset(8, -2, -3),
                pos.offset(-8, -2, 3),
                pos.offset(-8, -2, -3),
                pos.offset(3, -2, 8),
                pos.offset(-3, -2, 8),
                pos.offset(3, -2, -8),
                pos.offset(-3, -2, -8)
        };
        for (BlockPos i : coordinates) {
            if (level != null && level.getBlockState(i).getBlock() == GTLBlocks.POWER_CORE.get()) {
                BlockPos[] coordinatess = new BlockPos[] {
                        i.offset(3, 2, 0),
                        i.offset(-3, 2, 0),
                        i.offset(0, 2, 3),
                        i.offset(0, 2, -3)
                };
                for (BlockPos j : coordinatess) {
                    RecipeLogic logic = GTCapabilityHelper.getRecipeLogic(level, j, null);
                    if (logic != null && logic.getMachine().getDefinition() == AdvancedMultiBlockMachine.SPACE_ELEVATOR) {
                        if (logic.isWorking() && logic.getProgress() > 80) {
                            this.SpaceElevatorTier = ((SpaceElevatorMachine) logic.machine).getTier() - 7;
                            this.ModuleTier = ((SpaceElevatorMachine) logic.machine).getCasingTier();
                        } else if (!logic.isWorking()) {
                            this.SpaceElevatorTier = 0;
                            this.ModuleTier = 0;
                        }
                    }
                }
            }
        }
    }

    public static boolean beforeWorking(IRecipeLogicMachine machine, @NotNull GTRecipe recipe) {
        if (machine instanceof AdvancedSpaceElevatorModuleMachine spaceElevatorModuleMachine) {
            spaceElevatorModuleMachine.getSpaceElevatorTier();
            if (spaceElevatorModuleMachine.SpaceElevatorTier < 1) return false;
            return !spaceElevatorModuleMachine.SEPMTier || recipe.data.getInt("SEPMTier") <= spaceElevatorModuleMachine.ModuleTier;
        }
        return false;
    }

    public boolean onWorking() {
        boolean value = super.onWorking();
        if (this.getOffsetTimer() % 10L == 0L) {
            this.getSpaceElevatorTier();
            if (this.SpaceElevatorTier < 1) {
                this.getRecipeLogic().interruptRecipe();
                return false;
            }
        }
        return value;
    }

    public void addDisplayText(@NotNull List<Component> textList) {
        super.addDisplayText(textList);
        if (this.isFormed) {
            if (this.getOffsetTimer() % 10L == 0L) {
                this.getSpaceElevatorTier();
            }
            textList.add(Component.translatable("gtceu.multiblock.parallel",
                    Component.translatable(FormattingUtil.formatNumbers(getMaxParallel())).withStyle(ChatFormatting.DARK_PURPLE))
                    .withStyle(ChatFormatting.GRAY));
            textList.add(Component.translatable((this.SpaceElevatorTier < 1 ? "未" : "已") + "连接正在运行的太空电梯"));
        }
    }

    public int getMaxParallel() {
        return (2 << 3) << (this.ModuleTier - 1);
    }
}

File: src/main/java/com/gtladd/gtladditions/common/machine/muiltblock/controller/AdvancedSpaceElevatorModuleMachine.kt
package com.gtladd.gtladditions.common.machine.muiltblock.controller

import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.TickableSubscription
import com.gregtechceu.gtceu.api.machine.feature.IMachineLife
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic
import com.gtladd.gtladditions.api.machine.logic.GTLAddMultipleRecipesLogic
import com.gtladd.gtladditions.api.machine.multiblock.GTLAddWorkableElectricMultipleRecipesMachine
import com.lowdragmc.lowdraglib.syncdata.annotation.DescSynced
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import org.gtlcore.gtlcore.api.machine.multiblock.IModularMachineModule
import org.gtlcore.gtlcore.common.data.GTLBlocks
import org.gtlcore.gtlcore.common.machine.multiblock.electric.SpaceElevatorMachine
import org.gtlcore.gtlcore.utils.MachineUtil
import kotlin.math.pow

class AdvancedSpaceElevatorModuleMachine(holder: IMachineBlockEntity) :
    GTLAddWorkableElectricMultipleRecipesMachine(holder),
    IModularMachineModule<SpaceElevatorMachine, AdvancedSpaceElevatorModuleMachine>, IMachineLife {
    companion object {
        val MANAGED_FIELD_HOLDER: ManagedFieldHolder =
            ManagedFieldHolder(
                AdvancedSpaceElevatorModuleMachine::class.java,
                GTLAddWorkableElectricMultipleRecipesMachine.MANAGED_FIELD_HOLDER
            )
    }

    @field:DescSynced
    private var spaceElevatorTier = 0
    private var moduleTier = 0

    @field:Persisted
    private var hostPosition: BlockPos? = null
    private var host: SpaceElevatorMachine? = null

    override fun getHost(): SpaceElevatorMachine? {
        return host
    }

    override fun setHost(host: SpaceElevatorMachine?) {
        this.host = host
    }

    override fun getHostType(): Class<SpaceElevatorMachine> {
        return SpaceElevatorMachine::class.java
    }

    override fun getHostPosition(): BlockPos? {
        return hostPosition
    }

    override fun setHostPosition(pos: BlockPos?) {
        this.hostPosition = pos
    }

    override fun getFieldHolder(): ManagedFieldHolder {
        return MANAGED_FIELD_HOLDER
    }

    // ========================================
    // Update subscription for 80tick limit
    // ========================================

    private var updateSubs: TickableSubscription? = null
    private var remainingTicks: Int = 0

    private fun updateSubscription() {
        if (--this.remainingTicks > 0) {
            updateSubs = subscribeServerTick(updateSubs) { updateSubscription() }
        } else {
            update()
            updateSubs?.unsubscribe()
            updateSubs = null
        }
    }

    private fun update() {
        getSpaceElevatorTier()
        recipeLogic.updateTickSubscription()
    }

    private fun scheduleUpdate() {
        updateSubs?.unsubscribe()
        remainingTicks = 90
        updateSubscription()
    }

    // ========================================
    // Elevator connection
    // ========================================

    override fun onConnected(host: SpaceElevatorMachine) {
        if (host.recipeLogic.progress > 80) update() else scheduleUpdate()
    }

    override fun getHostScanPositions(): Array<out BlockPos?>? {
        (level as? ServerLevel)?.let { serverLevel ->
            val pos = getPos()
            val coordinates: Array<BlockPos> = arrayOf(
                pos.offset(8, -2, 3),
                pos.offset(8, -2, -3),
                pos.offset(-8, -2, 3),
                pos.offset(-8, -2, -3),
                pos.offset(3, -2, 8),
                pos.offset(-3, -2, 8),
                pos.offset(3, -2, -8),
                pos.offset(-3, -2, -8)
            )

            for (i in coordinates) {
                if (serverLevel.getBlockState(i).block === GTLBlocks.POWER_CORE.get()) {
                    return arrayOf(
                        i.offset(3, 2, 0),
                        i.offset(-3, 2, 0),
                        i.offset(0, 2, 3),
                        i.offset(0, 2, -3)
                    )
                }
            }
        }
        return MachineUtil.EMPTY_POS_ARRAY
    }

    private fun getSpaceElevatorTier() {
        host?.let { controller ->
            val logic = controller.getRecipeLogic()
            if (logic.isWorking && logic.progress > 80) {
                spaceElevatorTier = controller.tier - 7
                moduleTier = controller.casingTier
            } else {
                spaceElevatorTier = 0
                moduleTier = 0
            }
            return
        }
        spaceElevatorTier = 0
        moduleTier = 0
    }

    override fun createRecipeLogic(vararg args: Any): RecipeLogic {
        return GTLAddMultipleRecipesLogic(
            this
        ) { machine -> (machine as AdvancedSpaceElevatorModuleMachine).host != null }
    }

    override fun onStructureFormed() {
        super.onStructureFormed()
        if (!findAndConnectToHost()) {
            removeFromHost(this.host)
        }
    }

    override fun onStructureInvalid() {
        super.onStructureInvalid()
        removeFromHost(this.host)
    }

    override fun onPartUnload() {
        super.onPartUnload()
        removeFromHost(this.host)
    }

    override fun onMachineRemoved() {
        removeFromHost(this.host)
    }

    override fun onWorking(): Boolean {
        val value = super.onWorking()
        if (this.offsetTimer % 10L == 0L) {
            this.getSpaceElevatorTier()
            if (this.spaceElevatorTier < 1) getRecipeLogic().setProgress(0)
        }
        return value
    }

    override fun addDisplayText(textList: MutableList<Component?>) {
        super.addDisplayText(textList)
        if (this.isFormed) {
            if (this.offsetTimer % 10L == 0L) this.getSpaceElevatorTier()
            textList.add(Component.translatable((if (this.spaceElevatorTier < 1) "未" else "已") + "连接正在运行的太空电梯"))
        }
    }

    override fun getMaxParallel(): Int {
        return 8.0.pow((this.moduleTier - 1).toDouble()).toInt()
    }
}
